sympy/functions/combinatorial/factorials.py

```diff
@@ -480,37 +481,38 @@ def eval(cls, x, k):
                         else:
                             return S.Infinity
                     else:
-                        try:
-                            F, opt = poly_from_expr(x)
-                        except PolificationFailed:
-                            return reduce(lambda r, i: r*(x + i), range(0, int(k)), 1)
-                        if len(opt.gens) > 1 or F.degree() <= 1:
-                            return reduce(lambda r, i: r*(x + i), range(0, int(k)), 1)
+                        if isinstance(x, Poly):
+                            gens = x.gens
+                            if len(gens) > 1 or x.degree() <= 1:
```

Comment:
I was just going with what was already here. I'm not really sure if this is how it should work.